src/number_theory/modular_forms/congruence_subgroups.lean

```diff
@@ -0,0 +1,233 @@
+/-
+Copyright (c) 2022 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import linear_algebra.special_linear_group
+import data.zmod.basic
+import group_theory.subgroup.pointwise
+import group_theory.group_action.conj_act
+/-!
+# Congruence subgroups
+
+This defines congruence subgroups of `SL(2,ℤ)` such as `Γ(N)`, `Γ₀(N)` and `Γ₁(N)` for `N` a
+natural number.
+
+It also contains basic results about congruence subgroups.
+
+-/
+
+local notation `SL(` n `, ` R `)`:= matrix.special_linear_group (fin n) R
+
+local attribute [-instance] matrix.special_linear_group.has_coe_to_fun
+
+local prefix `↑ₘ`:1024 := @coe _ (matrix (fin 2) (fin 2) _) _
+
+open matrix.special_linear_group matrix
+
+variable (N : ℕ)
+
+local notation `SLMOD(`N`)`  := @matrix.special_linear_group.map (fin 2) _ _ _ _ _ _
+  (int.cast_ring_hom (zmod N))
+
+@[simp]
+lemma SL_reduction_mod_hom_val (N : ℕ) (γ : SL(2, ℤ)) : ∀ (i j : fin 2),
+  ((SLMOD(N) γ) : (matrix (fin 2) (fin 2) (zmod N))) i j =
+  (((↑ₘγ i j) : ℤ) : zmod N) := λ i j, rfl
+
+/--The full level `N` congruence subgroup of `SL(2,ℤ)` of matrices that reduce to the identity
+modulo `N`.-/
+def Gamma (N : ℕ) : subgroup SL(2, ℤ) := (SLMOD(N)).ker
+
+lemma Gamma_mem' (N : ℕ) (γ : SL(2, ℤ)) : γ ∈ (Gamma N) ↔ (SLMOD(N) γ) = 1 := iff.rfl
+
+@[simp]
+lemma Gamma_mem (N : ℕ) (γ : SL(2, ℤ)) : γ ∈ (Gamma N) ↔ (((↑ₘγ 0 0) : ℤ) : zmod N) = 1 ∧
+  (((↑ₘγ 0 1) : ℤ) : zmod N) = 0 ∧ (((↑ₘγ 1 0) : ℤ) : zmod N) = 0 ∧
+  (((↑ₘγ 1 1) : ℤ) : zmod N) = 1 :=
+begin
+  rw Gamma_mem',
+  split,
+  { intro h,
+    simp [←(SL_reduction_mod_hom_val N γ), h] },
+  { intro h,
+    ext,
+    rw SL_reduction_mod_hom_val N γ,
+    fin_cases i; fin_cases j,
+    all_goals {simp_rw h, refl} }
+end
+
+lemma Gamma_normal (N : ℕ) : subgroup.normal (Gamma N) := (SLMOD(N)).normal_ker
+
+lemma Gamma_one_top : (Gamma 1) = ⊤ :=
```

Comment:
```suggestion
lemma Gamma_one_top : Gamma 1 = ⊤ :=
```
There are a lot of extra brackets in this file!

Reply:
Ok hopefully I got them all...